File: backend/src/main/java/com/pairingplanet/pairing_planet/dto/recipe/UpdateRecipeRequestDto.java
package com.pairingplanet.pairing_planet.dto.recipe;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Size;

import java.util.List;
import java.util.UUID;

/**
 * Request DTO for updating an existing recipe.
 * Only the recipe creator can update, and only if there are no child variants or logs.
 */
public record UpdateRecipeRequestDto(
        @NotBlank(message = "제목은 필수입니다")
        @Size(min = 2, message = "제목은 최소 2자 이상이어야 합니다")
        String title,
        String description,
        String culinaryLocale,
        @NotEmpty(message = "재료는 최소 1개 이상 필요합니다")
        List<IngredientDto> ingredients,
        @NotEmpty(message = "조리 단계는 최소 1개 이상 필요합니다")
        List<StepDto> steps,
        @NotEmpty(message = "완성 사진은 최소 1장 이상 필요합니다")
        List<UUID> imagePublicIds,
        List<String> hashtags,
        // Servings (default: 2)
        Integer servings,
        // Cooking time range (e.g., "UNDER_15_MIN", "MIN_15_TO_30", etc.)
        String cookingTimeRange
) {}

File: frontend_mobile/lib/data/models/recipe/update_recipe_request_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pairing_planet2_frontend/domain/entities/recipe/update_recipe_request.dart';
import 'ingredient_dto.dart';
import 'step_dto.dart';

part 'update_recipe_request_dto.g.dart';

/// DTO for updating an existing recipe.
@JsonSerializable()
class UpdateRecipeRequestDto {
  final String title;
  final String? description;
  final String? culinaryLocale;
  final List<IngredientDto> ingredients;
  final List<StepDto> steps;
  final List<String> imagePublicIds;
  final List<String>? hashtags;

  UpdateRecipeRequestDto({
    required this.title,
    this.description,
    this.culinaryLocale,
    required this.ingredients,
    required this.steps,
    required this.imagePublicIds,
    this.hashtags,
  });

  factory UpdateRecipeRequestDto.fromJson(Map<String, dynamic> json) =>
      _$UpdateRecipeRequestDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UpdateRecipeRequestDtoToJson(this);

  factory UpdateRecipeRequestDto.fromEntity(UpdateRecipeRequest request) {
    return UpdateRecipeRequestDto(
      title: request.title,
      description: request.description,
      culinaryLocale: request.culinaryLocale,
      ingredients:
          request.ingredients.map((e) => IngredientDto.fromEntity(e)).toList(),
      steps: request.steps.map((e) => StepDto.fromEntity(e)).toList(),
      imagePublicIds: request.imagePublicIds,
      hashtags: request.hashtags,
    );
  }
}
